src/Compilers/CSharp/Portable/BoundTree/BoundDecisionDagNode.cs

```diff
@@ -46,5 +51,104 @@ public override int GetHashCode()
                     throw ExceptionUtilities.UnexpectedValue(this);
             }
         }
+
+#if DEBUG
+        private int _id;
+        private bool _idWasRead;
+
+        public int Id
+        {
+            get
+            {
+                _idWasRead = true;
+                return _id;
+            }
+            internal set
+            {
+                Debug.Assert(!_idWasRead, "Id was set after reading it");
+                _id = value;
+            }
+        }
+
+        internal new string GetDebuggerDisplay()
+        {
+            var pooledBuilder = PooledStringBuilder.GetInstance();
+            var builder = pooledBuilder.Builder;
+            builder.Append($"[{this.Id}]: ");
+            switch (this)
+            {
+                case BoundTestDecisionDagNode node:
+                    builder.Append($"{dumpDagTest(node.Test)} ");
+                    builder.Append(node.WhenTrue != null
+                        ? $"? [{node.WhenTrue.Id}] "
+                        : "? <unreachable> ");
+
+                    builder.Append(node.WhenFalse != null
+                        ? $": [{node.WhenFalse.Id}]"
+                        : ": <unreachable>");
+                    break;
+                case BoundEvaluationDecisionDagNode node:
+                    builder.Append($"{dumpDagTest(node.Evaluation)}; ");
+                    builder.Append(node.Next != null
+                        ? $"[{node.Next.Id}]"
+                        : "<unreachable>");
+                    break;
+                case BoundWhenDecisionDagNode node:
+                    builder.Append("when ");
+                    builder.Append(node.WhenExpression is { } when
+                        ? $"({when.Syntax}) "
+                        : "<true> ");
+
+                    builder.Append(node.WhenTrue != null
+                        ? $"? [{node.WhenTrue.Id}] "
+                        : "? <unreachable> ");
+
+                    builder.Append(node.WhenFalse != null
+                        ? $": [{node.WhenFalse.Id}]"
+                        : ": <unreachable>");
+
+                    break;
+                case BoundLeafDecisionDagNode node:
+                    builder.Append($"leaf {node.Label.Name}");
+                    break;
+                default:
+                    throw ExceptionUtilities.UnexpectedValue(this);
+            }
+            builder.AppendLine();
```

Comment:
nit: Consider shifting the newline to BoundDecisionDag.Dump instead

Reply:
Done